pkg/provisioners/message_dispatcher.go

```diff
@@ -223,3 +224,41 @@ func (d *MessageDispatcher) resolveURL(destination string, defaultNamespace stri
 		Path:   "/",
 	}
 }
+
+// ComputeRoutes sends the message to the router.
+func (d *MessageDispatcher) ComputeRoutes(message *Message, router string, defaults DispatchDefaults) ([]string, error) {
+	routerURL := d.resolveURL(router, defaults.Namespace)
+	req, err := http.NewRequest(http.MethodPost, routerURL.String(), bytes.NewReader(message.Payload))
+	if err != nil {
+		return nil, fmt.Errorf("unable to create request %v", err)
+	}
+	req.Header = d.toHTTPHeaders(message.Headers)
+
+	res, err := d.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	if res == nil {
+		// I don't think this is actually reachable with http.Client.Do(), but just to be sure we
+		// check anyway.
+		return nil, errors.New("non-error nil result from http.Client.Do()")
+	}
+	defer res.Body.Close()
+	if isFailure(res.StatusCode) {
+		// reject non-successful responses
+		return nil, fmt.Errorf("unexpected HTTP response, expected 2xx, got %d", res.StatusCode)
+	}
+
+	payload, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to read response %v", err)
+	}
+
+	var routes []string
+	err = json.Unmarshal(payload, &routes)
```

Comment:
I'm uncomfortable with the implied contract here:

CloudEvent -> [ subscription_names ]

In particular, it feels like there's some spooky action-at-a-distance in that the subscription names are based on shared knowledge of apiserver state (which may be laggy), rather than information in the request itself.

I'd much prefer a protocol where the choices were included in the request payload, so that the function could be implemented without needing reference to a third system.

Reply:
There is no apiserver state involved here. The subscription names are user-specified and the router does not manipulate any apiserver objects.

However there is the issue that the router can become out-of-sync with the channel subscribers, eg when manually adding `Subscription`. One solution is to provide two channel modes, pubsub vs routing. When a channel is in routing mode, only `Router` objects can manipulate the channel.